### ISSUE TITLE
Cockroach sometimes goes into an infinite loop and runs out of memory

### ISSUE THREAD

**Issue by arctica** [cockroachdb/cockroach#30363]
**BUG REPORT**

We are running a 4 node cluster in beta-production, all nodes are running version 2.0.5 currently.

The queries that the cluster is executing are pretty much the same apart from some small string parameters. The cluster runs fine for hours or even a few days but then sometimes seems to run into some kind of infinite loop which causes it to allocate all available RAM on the server, eventually swapping and getting killed by the OOM killer.

The 4 nodes: master, node1, node2 and node3. In the latest instance, node3 seems to have run into the issue a few minutes before the other nodes started misbehaving.

It's a bit hard to debug because it happens seemingly randomly and when it happens it's not so easy to get any info out of the nodes when they are using all CPU and swapping. But below is a heap profile that I was able to obtain before a node ran into swap:

```
(pprof) top20
Showing nodes accounting for 32733.41MB, 97.25% of 33660.47MB total
Dropped 462 nodes (cum <= 168.30MB)
Showing top 20 nodes out of 40
      flat  flat%   sum%        cum   cum%
28925.71MB 85.93% 85.93% 28945.73MB 85.99%  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*mergeJoiner).nextRow
 2872.92MB  8.54% 94.47%  2872.92MB  8.54%  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*routerBase).fwdMetadata
  388.16MB  1.15% 95.62%   392.67MB  1.17%  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ProcOutputHelper).ProcessRow
  292.55MB  0.87% 96.49%   292.55MB  0.87%  github.com/cockroachdb/cockroach/pkg/storage/engine.copyFromSliceVector
  240.07MB  0.71% 97.20%   240.07MB  0.71%  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*routerOutput).popRowsLocked
       8MB 0.024% 97.23%   188.32MB  0.56%  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*clientStream).SendMsg
       5MB 0.015% 97.24%   174.25MB  0.52%  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.encode
    0.50MB 0.0015% 97.24%   422.05MB  1.25%  github.com/cockroachdb/cockroach/pkg/storage.(*Store).Send
    0.50MB 0.0015% 97.25%   475.69MB  1.41%  github.com/cockroachdb/cockroach/pkg/kv.(*grpcTransport).send
         0     0% 97.25%   475.69MB  1.41%  github.com/cockroachdb/cockroach/pkg/kv.(*grpcTransport).SendNext.func1
         0     0% 97.25%   475.19MB  1.41%  github.com/cockroachdb/cockroach/pkg/kv.(*grpcTransport).send.func1
         0     0% 97.25%   422.05MB  1.25%  github.com/cockroachdb/cockroach/pkg/server.(*Node).Batch
         0     0% 97.25%   422.05MB  1.25%  github.com/cockroachdb/cockroach/pkg/server.(*Node).batchInternal
         0     0% 97.25%   422.05MB  1.25%  github.com/cockroachdb/cockroach/pkg/server.(*Node).batchInternal.func1
         0     0% 97.25%   186.32MB  0.55%  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*distSQLFlowStreamClient).Send
         0     0% 97.25%  3055.23MB  9.08%  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*hashRouter).Push
         0     0% 97.25%   182.31MB  0.54%  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*memRowContainer).AddRow
         0     0% 97.25% 28950.23MB 86.01%  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*mergeJoiner).Next
         0     0% 97.25% 31823.16MB 94.54%  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*mergeJoiner).Run
         0     0% 97.25%   188.31MB  0.56%  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*outbox).addRow
```

A `SHOW queries;` didn't show anything weird running. The biggest DB is 1.2GB so even fetching all data from it shouldn't allocate so much RAM.

I'm attaching two logs. One from node3 and one from master. From the runtime stats, one can clearly see when memory usage started to skyrocket so that's around the time the problem started. The one thing that stood out to me from the logs of node3 is the following line:

`I180918 12:00:11.563659 7380819 sql/distsqlrun/hashjoiner.go:217  [n3,HashJoiner] buffer phase error [n3] communication error: rpc error: code = Canceled desc = context canceled`

While there are a lot more entries in the log of "master", I spotted the exact same error in there and it is also in between the last stable "runtime stats" entry and the one where an immediate increase in RSS usage is apparent.

```
I180918 12:03:32.406053 248 server/status/runtime.go:219  [n1] runtime stats: 19 GiB RSS, 320 goroutines, 202 MiB/340 MiB/623 MiB GO alloc/idle/total, 16 GiB/19 GiB CGO alloc/total, 403.70cgo/sec, 0.12/0.02 %(u/s)time, 0.00 %gc (1x)
...
...
I180918 12:03:52.406170 248 server/status/runtime.go:219  [n1] runtime stats: 19 GiB RSS, 383 goroutines, 298 MiB/246 MiB/623 MiB GO alloc/idle/total, 16 GiB/19 GiB CGO alloc/total, 99.81cgo/sec, 0.04/0.02 %(u/s)time, 0.00 %gc (0x)
E180918 12:03:52.706742 27184225 sql/distsqlrun/server.go:476  [n1] query execution canceled
E180918 12:03:52.706750 27184224 sql/distsqlrun/server.go:476  [n1] query execution canceled
E180918 12:03:52.706743 27184318 sql/distsqlrun/server.go:476  [n1] query execution canceled
I180918 12:03:52.706949 27184355 sql/distsqlrun/hashjoiner.go:217  [n1,HashJoiner] buffer phase error [n1] communication error: rpc error: code = Canceled desc = context canceled
...
I180918 12:04:02.405186 248 server/status/runtime.go:219  [n1] runtime stats: 22 GiB RSS, 246 goroutines, 3.7 GiB/544 MiB/4.5 GiB GO alloc/idle/total, 16 GiB/19 GiB CGO alloc/total, 551.84cgo/sec, 3.23/0.22 %(u/s)time, 0.01 %gc (9x)
...
...
I180918 12:04:15.323249 248 server/status/runtime.go:219  [n1] runtime stats: 26 GiB RSS, 251 goroutines, 8.0 GiB/287 MiB/8.7 GiB GO alloc/idle/total, 16 GiB/19 GiB CGO alloc/total, 370.03cgo/sec, 3.28/0.26 %(u/s)time, 0.01 %gc (3x)
```

The fact that the errors in the log mention a HashJoiner and the heap profile showing all memory being allocated in a mergeJoiner maybe can be a good hint as to where the bug might be. In "show queries;", we've actually seen a few of our standard queries running which use a JOIN though they get executed all the time so it's not as easy as executing one and reproducing the bug.

BTW: Even after restarting 3 out of the 4 nodes, running a query seems to be stuck (or maybe takes extremely long).

[node3.log](https://github.com/cockroachdb/cockroach/files/2393224/node3.log)
[master.log](https://github.com/cockroachdb/cockroach/files/2393225/master.log)

**Comment by jordanlewis** [comment]
@solongordon could this be the metadata forwarding bug you fixed recently?

**Comment by solongordon** [comment]
I don't think so. It's reminiscent, but that bug was introduced in 2.1.

**Comment by jordanlewis** [comment]
@arctica could you attach the heap dump if possible?

**Comment by jordanlewis** [comment]
If not, outputting `top` from the profile you collected after running `lines` would be sufficient. Also please include the release you're running.

**Comment by arctica** [comment]
Build info:

`I180918 12:57:35.852055 1 util/log/clog.go:1134  [config] binary: CockroachDB CCL v2.0.5 (x86_64-unknown-linux-gnu, built 2018/08/13 17:59:42, go1.10)`

pprof `top` output after `lines`:

```
(pprof) lines
(pprof) top
Showing nodes accounting for 31.95GB, 97.20% of 32.87GB total
Dropped 574 nodes (cum <= 0.16GB)
Showing top 10 nodes out of 38
      flat  flat%   sum%        cum   cum%
   28.25GB 85.93% 85.93%    28.25GB 85.93%  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*mergeJoiner).nextRow /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/mergejoiner.go:209
    2.81GB  8.54% 94.47%     2.81GB  8.54%  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*routerBase).fwdMetadata /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/routers.go:103
    0.38GB  1.15% 95.62%     0.38GB  1.15%  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ProcOutputHelper).ProcessRow /go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlbase/encoded_datum.go:467
    0.29GB  0.87% 96.49%     0.29GB  0.87%  github.com/cockroachdb/cockroach/pkg/storage/engine.copyFromSliceVector /go/src/github.com/cockroachdb/cockroach/pkg/storage/engine/rocksdb.go:2173
    0.23GB  0.71% 97.20%     0.23GB  0.71%  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*routerOutput).popRowsLocked /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/routers.go:134
         0     0% 97.20%     0.46GB  1.41%  github.com/cockroachdb/cockroach/pkg/kv.(*grpcTransport).SendNext.func1 /go/src/github.com/cockroachdb/cockroach/pkg/kv/transport.go:223
         0     0% 97.20%     0.46GB  1.41%  github.com/cockroachdb/cockroach/pkg/kv.(*grpcTransport).send /go/src/github.com/cockroachdb/cockroach/pkg/kv/transport.go:279
         0     0% 97.20%     0.38GB  1.17%  github.com/cockroachdb/cockroach/pkg/kv.(*grpcTransport).send.func1 /go/src/github.com/cockroachdb/cockroach/pkg/kv/transport.go:248
         0     0% 97.20%     0.41GB  1.25%  github.com/cockroachdb/cockroach/pkg/server.(*Node).Batch /go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:939
         0     0% 97.20%     0.41GB  1.25%  github.com/cockroachdb/cockroach/pkg/server.(*Node).batchInternal /go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:898
```

Attaching also the full pprof dump.
[pprof.cockroach.alloc_objects.alloc_space.inuse_objects.inuse_space.001.pb.gz](https://github.com/cockroachdb/cockroach/files/2393370/pprof.cockroach.alloc_objects.alloc_space.inuse_objects.inuse_space.001.pb.gz)

Note that the dump is not from the exact same instance of the problem as the logs are but from an earlier one. I don't think it makes any difference though.

**Comment by arctica** [comment]
So the heap profile links to mergeJoiner.go line 209: https://github.com/cockroachdb/cockroach/blob/v2.0.5/pkg/sql/distsqlrun/mergejoiner.go#L206

Which is code that checks the cancellation of a query as far as I can see. That fits perfectly with the error in the logs just before the nodes go funky.

**Comment by solongordon** [comment]
Yes, I think I see the problem. The merge joiner does not close itself after encountering cancellation, so presumably its consumer keeps calling `Next()` and generating the same "query execution canceled" error over and over.

This code has been refactored for 2.1 so it should no longer be an issue, but I'll put together a patch for 2.0.5.

**Comment by solongordon** [comment]
Thank you for the detailed report! I'm going to close this issue but please do re-open if you continue to see the issue in v2.1. I will warn you that a similar issue was recently discovered in v2.1 (see #29978). Looks like the fix did not make it into v2.1.0-beta.20180917, but it will be present in the next beta.
